skgstat/MetricSpace.py

```diff
@@ -243,3 +255,107 @@ def dists(self):
 
         # return
         return self._dists
+
+class ProbabalisticMetricSpace(MetricSpace):
+    """Like MetricSpace but samples the distance pairs only returning a
+       `samples` sized subset. `samples` can either be a fraction of
+       the total number of pairs (float < 1), or an integer count.
+    """
+    
+    def __init__(self, coords, dist_metric="euclidean", max_dist=None, samples=0.5):
+        """ProbabalisticMetricSpace class
+
+        Parameters
+        ----------
+        coords : numpy.ndarray
+            Coordinate array of shape (Npoints, Ndim)
+        dist_metric : str
+            Distance metric names as used by scipy.spatial.distance.pdist
+        max_dist : float
+            Maximum distance between points after which the distance
+            is considered infinite and not calculated.
+        samples : float, int
+            Number of samples (int) or fraction of coords to sample (float < 1).
+        """
+        
+        self.coords = coords.copy()
+        self.dist_metric = dist_metric
+        self.max_dist = max_dist
+        self.samples = samples
+
+        self._lidx = None
+        self._ridx = None
+        self._ltree = None
+        self._rtree = None
+        self._dists = None
+        # Do a very quick check to see throw exceptions if self.dist_metric is invalid...
+        pdist(self.coords[:1,:], metric=self.dist_metric)
+
+    @property
+    def sample_count(self):
+        if isinstance(self.samples, int):
+            return self.samples
+        return int(self.samples * len(self.coords))
+        
+    @property
+    def lidx(self):
+        """The sampled indices into `self.coords` for the left sample."""
+        if self._lidx is None:
+            self._lidx = np.random.choice(len(self.coords), size=self.sample_count, replace=False)
+        return self._lidx
+    
+    @property
+    def ridx(self):
+        """The sampled indices into `self.coords` for the right sample."""
+        if self._ridx is None:
+            self._ridx = np.random.choice(len(self.coords), size=self.sample_count, replace=False)
```

Comment:
same here

Reply:
Wait, so np.random.seed(42) does not apply "globally" to all numpy random operations after that call? That's what I did in the unittest. Maybe I've just misunderstood np.random a bit, and my code only works by accident?

Reply:
No, I think I was misconcepted here. The unittests work fine, so I guess that's fine. 
The reason why I like to have something like `ProbabilisticMetricSpace(..., seed=1234)` is that the seed appears as an argument to the user indicating that the class can somewhat be seeded. 
That is basically the main reason why I would like to have such an argument... 

What do you think?

Reply:
Hm, maybe it _is_ better with a parameter after all, as np.random.choice is only called when a property is first evaluated so it's non-trivial to know what the seed actually is at that point?

```
s = ProbabilisticMetricSpace()
np.random.seed(42)
s.lidx, s.ridx

```

isn't a particularly good API ;)